File: services/api/app/schemas/knowledge.py
from datetime import datetime
from enum import StrEnum
from typing import Literal
from uuid import UUID

from pydantic import Field, field_validator

from app.schemas.reports import StrictSchema


class KnowledgeReviewStatus(StrEnum):
    DRAFT = "draft"
    DEMO = "demo"
    APPROVED = "approved"
    REJECTED = "rejected"


class KnowledgeReviewHistory(StrictSchema):
    old_status: KnowledgeReviewStatus | None
    new_status: KnowledgeReviewStatus
    actor_type: str
    actor_display_name: str | None
    reason: str
    created_at: datetime


class KnowledgePreview(StrictSchema):
    source_type: str
    character_count: int
    chunk_count: int
    preview: str


class KnowledgeDocument(StrictSchema):
    id: UUID
    title: str
    category: str | None
    source_type: str
    administrative_unit_id: UUID
    is_mandatory: bool
    is_active: bool
    processing_status: str
    failure_message: str | None
    review_status: KnowledgeReviewStatus
    reviewer_display_name: str | None = None
    reviewed_at: datetime | None = None
    review_reason: str | None = None
    allowed_review_transitions: list[KnowledgeReviewStatus] = Field(
        default_factory=list
    )
    created_at: datetime
    updated_at: datetime


class KnowledgeDocumentList(StrictSchema):
    items: list[KnowledgeDocument]


class KnowledgeDocumentDetail(KnowledgeDocument):
    content: str
    service_key: str | None
    review_history: list[KnowledgeReviewHistory] = Field(default_factory=list)


class KnowledgeUpdate(StrictSchema):
    title: str | None = Field(default=None, min_length=1, max_length=300)
    category: str | None = Field(default=None, max_length=100)
    is_active: bool | None = None
    is_mandatory: bool | None = None
    content: str | None = Field(default=None, min_length=1)
    service_key: str | None = Field(
        default=None,
        pattern=r"^[a-z0-9_-]+$",
    )

    @field_validator("title", "content")
    @classmethod
    def reject_blank(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        if not normalized:
            raise ValueError("value must not be blank")
        return normalized


class KnowledgeReviewUpdate(StrictSchema):
    status: Literal[
        KnowledgeReviewStatus.APPROVED,
        KnowledgeReviewStatus.REJECTED,
    ]
    reason: str = Field(min_length=1, max_length=1000)

    @field_validator("reason")
    @classmethod
    def reject_blank_reason(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("reason must not be blank")
        return normalized


class EmbeddingChunk(StrictSchema):
    chunk_id: UUID
    content: str
    checksum: str


class EmbeddingJob(StrictSchema):
    document_id: UUID
    chunks: list[EmbeddingChunk]


class EmbeddingValue(StrictSchema):
    chunk_id: UUID
    embedding: list[float] = Field(min_length=768, max_length=768)


class EmbeddingComplete(StrictSchema):
    embeddings: list[EmbeddingValue]


class EmbeddingFailure(StrictSchema):
    message: str = Field(min_length=1, max_length=500)

File: apps/dashboard/lib/knowledge.ts
export type KnowledgeReviewStatus = "draft" | "demo" | "approved" | "rejected";

export type KnowledgeReviewHistory = {
  old_status: KnowledgeReviewStatus | null;
  new_status: KnowledgeReviewStatus;
  actor_type: string;
  actor_display_name: string | null;
  reason: string;
  created_at: string;
};

export type KnowledgeDocument = {
  id: string;
  title: string;
  category: string | null;
  source_type: string;
  administrative_unit_id: string;
  is_mandatory: boolean;
  is_active: boolean;
  processing_status: "pending" | "processing" | "ready" | "failed";
  failure_message: string | null;
  review_status: KnowledgeReviewStatus;
  reviewer_display_name: string | null;
  reviewed_at: string | null;
  review_reason: string | null;
  allowed_review_transitions: KnowledgeReviewStatus[];
  created_at: string;
  updated_at: string;
};

export type KnowledgeDocumentDetail = KnowledgeDocument & {
  content: string;
  service_key: string | null;
  review_history: KnowledgeReviewHistory[];
};

export class KnowledgeApiError extends Error {
  readonly status: number;

  constructor(status: number) {
    super("Knowledge API request failed");
    this.name = "KnowledgeApiError";
    this.status = status;
  }
}

function endpoint(path: string) {
  const baseUrl = process.env.NEXT_PUBLIC_API_URL?.trim();
  if (!baseUrl) throw new Error("NEXT_PUBLIC_API_URL is not configured");
  return new URL(path, `${baseUrl.replace(/\/+$/, "")}/`);
}

async function request<T>(path: string, token: string, init?: RequestInit): Promise<T> {
  const response = await fetch(endpoint(path), {
    ...init,
    cache: "no-store",
    signal: init?.signal ?? AbortSignal.timeout(init?.body instanceof FormData ? 60_000 : init?.method && init.method !== "GET" ? 30_000 : 15_000),
    headers: {
      ...init?.headers,
      Authorization: `Bearer ${token}`,
    },
  });
  if (!response.ok) throw new KnowledgeApiError(response.status);
  if (response.status === 204) return undefined as T;
  return response.json() as Promise<T>;
}

export function listKnowledgeDocuments(token: string) {
  return request<{ items: KnowledgeDocument[] }>(
    "/api/v1/knowledge/documents",
    token,
  );
}

export function getKnowledgeDocument(id: string, token: string) {
  return request<KnowledgeDocumentDetail>(
    `/api/v1/knowledge/documents/${encodeURIComponent(id)}`,
    token,
  );
}

export function createKnowledgeDocument(input: FormData, token: string) {
  return request<KnowledgeDocument>("/api/v1/knowledge/documents", token, {
    method: "POST",
    body: input,
  });
}

export function updateKnowledgeDocument(
  id: string,
  input: Record<string, unknown>,
  token: string,
) {
  return request<KnowledgeDocumentDetail>(
    `/api/v1/knowledge/documents/${encodeURIComponent(id)}`,
    token,
    {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    },
  );
}

export function reviewKnowledgeDocument(
  id: string,
  input: { status: "approved" | "rejected"; reason: string },
  token: string,
) {
  return request<KnowledgeDocumentDetail>(
    `/api/v1/knowledge/documents/${encodeURIComponent(id)}/review`,
    token,
    {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    },
  );
}

export function deactivateKnowledgeDocument(id: string, token: string) {
  return request<void>(
    `/api/v1/knowledge/documents/${encodeURIComponent(id)}`,
    token,
    { method: "DELETE" },
  );
}
